Fatal error signing Mac App

**Describe the bug**

I am building a desktop application. The build works completely fine without signing. With signing, I get the following error:

```
Caused by: org.gradle.api.InvalidUserCodeException: Starting an external process '/usr/bin/security find-certificate -a -c Developer ID Application: <my dev id> <my-keychain-path>' during configuration time is unsupported.
```

**Affected platforms**

Desktop

**Versions**
- Kotlin version*:  1.8.20
- Compose Multiplatform version*: compiler version 1.4.5, plugin version 1.4.0
- OS version: Mac OSX 
- OS architecture (x86 or arm64): arm64
- JDK (for desktop issues): 17
- Gradle: 8.1

**To Reproduce**
1. Create a desktop project using the versions above. I suspect the gradle version is the most critical one.
2. In the buildscript, configure the mac app to be signed
3. Execute the build

**Expected behavior**

The build to succeed

**Workaround**

Use `--no-configuration-cache` in the gradle process. However, this prevents using the configuration cache.

**Additional context**

I suspect this issue is related to the latest Gradle version. Gradle has been tightening restrictions related to the configuration cache.

--- a/gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacSigner.kt
+++ b/gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacSigner.kt
@@ -55,28 +55,33 @@ internal class NoCertificateSigner(runTool: ExternalToolRunner) : MacSigner(runT
 internal class MacSignerImpl(
     override val settings: ValidatedMacOSSigningSettings,
     runTool: ExternalToolRunner
-) : MacSigner(runTool)  {
-    private lateinit var signKey: String
-
-    init {
-        runTool(
-            MacUtils.security,
-            args = listOfNotNull(
-                "find-certificate",
-                "-a",
-                "-c",
-                settings.fullDeveloperID,
-                settings.keychain?.absolutePath
-            ),
-            processStdout = { signKey = matchCertificates(it) }
-        )
-    }
+) : MacSigner(runTool) {
+    @Transient
+    private var signKeyValue: String? = null
 
     override fun sign(
         file: File,
         entitlements: File?,
         forceEntitlements: Boolean
     ) {
+        // sign key calculation is delayed to avoid
+        // creating an external process during the configuration
+        // phase, which became an error in Gradle 8.1
+        // https://github.com/JetBrains/compose-multiplatform/issues/3060
+        val signKey = signKeyValue ?: run {
+            runTool(
+                MacUtils.security,
+                args = listOfNotNull(
+                    "find-certificate",
+                    "-a",
+                    "-c",
+                    settings.fullDeveloperID,
+                    settings.keychain?.absolutePath
+                ),
+                processStdout = { signKeyValue = matchCertificates(it) }
+            )
+            signKeyValue!!
+        }
         runTool.unsign(file)
         runTool.sign(
             file = file,

--- a/gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/test/tests/integration/DesktopApplicationTest.kt
+++ b/gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/test/tests/integration/DesktopApplicationTest.kt
@@ -334,6 +334,15 @@ class DesktopApplicationTest : GradlePluginTestBase() {
     }
 
     @Test
+    fun testMacSignConfiguration() {
+        Assumptions.assumeTrue(currentOS == OS.MacOS)
+
+        with(testProject(TestProjects.macSign)) {
+            gradle("--dry-run", ":createDistributable")
+        }
+    }
+
+    @Test
     @Disabled
     // the test does not work on CI and locally unless test keychain is opened manually
     fun testMacSign() {

--- a/gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/test/tests/integration/GradlePluginTest.kt
+++ b/gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/test/tests/integration/GradlePluginTest.kt
@@ -20,7 +20,14 @@ import org.junit.jupiter.api.Test
 
 class GradlePluginTest : GradlePluginTestBase() {
     @Test
-    fun skikoWasm() = with(testProject(TestProjects.skikoWasm)) {
+    fun skikoWasm() = with(
+        testProject(
+            TestProjects.skikoWasm,
+            // configuration cache is disabled as a temporary workaround for KT-58057
+            // todo: enable once KT-58057 is fixed
+            testEnvironment = defaultTestEnvironment.copy(useGradleConfigurationCache = false)
+        )
+    ) {
         gradle(":build").checks {
             check.taskSuccessful(":unpackSkikoWasmRuntimeJs")
             check.taskSuccessful(":compileKotlinJs")

--- a/gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/test/utils/TestProject.kt
+++ b/gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/test/utils/TestProject.kt
@@ -19,6 +19,7 @@ data class TestEnvironment(
     val composeCompilerPlugin: String? = null,
     val composeCompilerArgs: String? = null,
     val composeVerbose: Boolean = true,
+    val useGradleConfigurationCache: Boolean = TestProperties.gradleConfigurationCache,
 ) {
     private val placeholders = linkedMapOf(
         "COMPOSE_GRADLE_PLUGIN_VERSION_PLACEHOLDER" to composeGradlePluginVersion,
@@ -76,7 +77,7 @@ class TestProject(
     }
 
     internal fun gradle(vararg args: String): BuildResult {
-        if (TestProperties.gradleConfigurationCache) {
+        if (testEnvironment.useGradleConfigurationCache) {
             if (GradleVersion.version(TestProperties.gradleVersionForTests).baseVersion < GradleVersion.version("8.0")) {
                 // Gradle 7.* does not use the configuration cache in the same build.
                 // In other words, if cache misses, Gradle performs configuration,
@@ -106,7 +107,7 @@ class TestProject(
     private fun gradleRunner(args: Array<out String>): GradleRunner {
         val allArgs = args.toMutableList()
         allArgs.addAll(additionalArgs)
-        if (TestProperties.gradleConfigurationCache) {
+        if (testEnvironment.useGradleConfigurationCache) {
             allArgs.add("--configuration-cache")
         }
 

--- a/gradle-plugins/gradle.properties
+++ b/gradle-plugins/gradle.properties
@@ -10,7 +10,7 @@ compose.tests.compiler.compatible.kotlin.version=1.8.20
 # The latest version of Kotlin compatible with compose.tests.compiler.version for JS target. Used only on CI.
 compose.tests.js.compiler.compatible.kotlin.version=1.8.20
 # __SUPPORTED_GRADLE_VERSIONS__
-compose.tests.gradle.versions=7.3.3, 8.0.2
+compose.tests.gradle.versions=7.3.3, 8.1
 
 # A version of Gradle plugin, that will be published,
 # unless overridden by COMPOSE_GRADLE_PLUGIN_VERSION env var.

